controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/WorkerPool.java

```diff
@@ -37,40 +42,49 @@
   // TODO: graceful shutdown.
   private static final Logger LOG = LoggerFactory.getLogger(WorkerPool.class);
   private final Semaphore idleWorkersSemaphore;
-  private final ConcurrentHashMap<String, FutureTask> runningTasks;
+  private final ConcurrentHashMap<String, Future> runningTasks;
   private final ExecutorService executorService;
+  private final TaskQueue taskQueue;
 
-  public WorkerPool(ExecutorService executorService, Semaphore idleWorkersSemaphore) {
+  public WorkerPool(ExecutorService executorService,
+                    Semaphore idleWorkersSemaphore,
+                    TaskQueue taskQueue) {
     this.executorService = executorService;
     this.idleWorkersSemaphore = idleWorkersSemaphore;
     this.runningTasks = new ConcurrentHashMap<>();
+    this.taskQueue = taskQueue;
   }
 
   /**
    * Assign a task to a worker thread.
    * @param task the task to execute
    * @throws Exception if there is a running task for the cluster.
    */
-  public boolean assignTask(TaskBase task) {
+  public synchronized boolean assignTask(TaskInternal task) throws Exception {
```

Comment:
There may be different threads calling #abortTask(), thus it's safer to use synchronized.